store/etcd/put.go

```diff
@@ -17,222 +14,146 @@ import (
-	sig, err := validateRuleset(path, ruleset)
-	if err != nil {
-		return nil, err
-	}
-
-	var entry store.RulesetEntry
+// Put stores the given rules under the rules tree. If no signature is found for the given path it returns an error.
+func (s *RulesetService) Put(ctx context.Context, path string, rules []*rule.Rule) (*regula.Ruleset, error) {
+	var ruleset *regula.Ruleset
+	var err error
 
 	txfn := func(stm concurrency.STM) error {
-		// generate a checksum from the ruleset for comparison purpose
-		h := md5.New()
-		err = json.NewEncoder(h).Encode(ruleset)
-		if err != nil {
-			return errors.Wrap(err, "failed to generate checksum")
-		}
-		checksum := string(h.Sum(nil))
-
-		// if nothing changed return latest ruleset
-		if stm.Get(s.checksumsPath(path)) == checksum {
-			v := stm.Get(stm.Get(s.latestRulesetPath(path)))
-
-			var pbrse pb.RulesetEntry
-			err = proto.Unmarshal([]byte(v), &pbrse)
-			if err != nil {
-				s.Logger.Debug().Err(err).Str("entry", v).Msg("put: entry unmarshalling failed")
-				return errors.Wrap(err, "failed to unmarshal entry")
-			}
-
-			s.Logger.Debug().Str("path", path).Msg("ruleset didn't change, returning without creating a new version")
-
-			entry.Path = pbrse.Path
-			entry.Version = pbrse.Version
-			entry.Ruleset = rulesetFromProtobuf(pbrse.Ruleset)
-
-			return store.ErrNotModified
-		}
-
-		// make sure signature didn't change
-		rawSig := stm.Get(s.signaturesPath(path))
-		if rawSig != "" {
-			var pbsig pb.Signature
-
-			err := proto.Unmarshal([]byte(rawSig), &pbsig)
-			if err != nil {
-				s.Logger.Debug().Err(err).Str("signature", rawSig).Msg("put: signature unmarshalling failed")
-				return errors.Wrap(err, "failed to decode ruleset signature")
-			}
-
-			err = compareSignature(signatureFromProtobuf(&pbsig), sig)
-			if err != nil {
-				return err
-			}
-		}
-
-		// if no signature found, create one
-		if rawSig == "" {
-			b, err := proto.Marshal(signatureToProtobuf(sig))
-			if err != nil {
-				return errors.Wrap(err, "failed to encode updated signature")
-			}
-
-			stm.Put(s.signaturesPath(path), string(b))
-		}
+		p := rulesPutter{s, stm}
+		ruleset, err = p.put(ctx, path, rules)
+		return err
+	}
 
-		// update checksum
-		stm.Put(s.checksumsPath(path), checksum)
+	_, err = concurrency.NewSTM(s.Client, txfn, concurrency.WithAbortContext(ctx))
+	if err != nil && err != store.ErrRulesetNotModified && !store.IsValidationError(err) {
+		return nil, errors.Wrap(err, "failed to put ruleset")
+	}
 
-		// create a new ruleset version
-		k, err := ksuid.NewRandom()
-		if err != nil {
-			return errors.Wrap(err, "failed to generate ruleset version")
-		}
-		version := k.String()
+	return ruleset, err
+}
 
-		err = s.putVersions(stm, path, version)
-		if err != nil {
-			return err
-		}
+// rulesPutter is responsible for validating and storing rules, updating checksums and other actions
+// that are required in order to add a new ruleset version correctly.
+type rulesPutter struct {
```

Comment:
This "Putter" wording amuses me every time I see it.  "Putter" is a real word in English, but it doesn't describe something that "puts".  Here are the definitions:
```
 putter (pŭtˈər)►

    n.
    A short golf club used for putting.
    n.
    A golfer who is putting.
    v.
    To occupy oneself in an aimless or ineffective manner.
```
Funnily enough, I don't think there is a noun for someone or something that "puts", so I've no better suggestion.  I'll just carry on being amused.

Reply:
It troubled me too! To be honest I wrote it expecting that you'd propose a better one if this one was too weird. But you failed. Shame on you.